src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -366,26 +365,70 @@ public virtual void Write(object value)
         /// </remarks>
         public virtual void WriteTo([NotNull] TextWriter writer, object value)
         {
-            if (value != null && value != HtmlString.Empty)
+            WriteTo(writer, HtmlEncoder, value, escapeQuotes: false);
+        }
+
+        /// <summary>
+        /// Writes the specified <paramref name="value"/> with HTML encoding to given <paramref name="writer"/>.
+        /// </summary>
+        /// <param name="writer">The <see cref="TextWriter"/> instance to write to.</param>
+        /// <param name="encoder">The <see cref="IHtmlEncoder"/> to use when encoding <paramref name="value"/>.</param>
+        /// <param name="value">The <see cref="object"/> to write.</param>
+        /// <param name="escapeQuotes">
+        /// If <c>true</c> escapes double quotes in a <paramref name="value"/> of type <see cref="HtmlString"/>.
+        /// Otherwise writes <see cref="HtmlString"/> values as-is.
+        /// </param>
+        /// <remarks>
+        /// <paramref name="value"/>s of type <see cref="HtmlString"/> are written without encoding and the
+        /// <see cref="HelperResult.WriteTo(TextWriter)"/> is invoked for <see cref="HelperResult"/> types.
+        /// For all other types, the encoded result of <see cref="object.ToString"/> is written to the
+        /// <paramref name="writer"/>.
+        /// </remarks>
+        public static void WriteTo(
+            [NotNull] TextWriter writer,
+            [NotNull] IHtmlEncoder encoder,
+            object value,
+            bool escapeQuotes)
+        {
+            if (value == null || value == HtmlString.Empty)
             {
-                var helperResult = value as HelperResult;
-                if (helperResult != null)
-                {
-                    helperResult.WriteTo(writer);
-                }
-                else
+                return;
+            }
+
+            var helperResult = value as HelperResult;
+            if (helperResult != null)
+            {
+                helperResult.WriteTo(writer);
+                return;
+            }
+
+            var htmlString = value as HtmlString;
+            if (htmlString != null)
+            {
+                if (escapeQuotes)
                 {
-                    var htmlString = value as HtmlString;
-                    if (htmlString != null)
+                    // In this case the text likely came directly from the Razor source. Since the original string is
+                    // an attribute value that may have been quoted with single quotes, must handle any double quotes
+                    // in the value. Writing the value out surrounded by double quotes.
+                    var stringValue = htmlString.ToString();
+                    if (stringValue.Contains("\""))
```

Comment:
`htmlString.ToString()` is very inefficient when the `HtmlString` has been instantiated using a `StringCollectionTextWriter`.  bad idea to move that into the mainline code path.

perhaps a `return` and fall-through would work.  will play a bit...